src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonImmutableDictionaryOfStringTValueConverter.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    internal sealed class JsonImmutableDictionaryOfStringTValueConverter<TCollection, TValue> : JsonDictionaryDefaultConverter<TCollection, TValue>
+        where TCollection : IReadOnlyDictionary<string, TValue>
+    {
+        protected override void Add(TValue value, JsonSerializerOptions options, ref ReadStack state)
+        {
+            Debug.Assert(state.Current.ReturnValue is Dictionary<string, TValue>);
+
+            string key = state.Current.JsonPropertyNameAsString!;
+            ((Dictionary<string, TValue>)state.Current.ReturnValue!)[key] = value;
+        }
+
+        internal override bool CanHaveIdMetadata => false;
+
+        protected override void CreateCollection(ref ReadStack state)
+        {
+            state.Current.ReturnValue = new Dictionary<string, TValue>();
+        }
+
+        protected override void ConvertCollection(ref ReadStack state, JsonSerializerOptions options)
+        {
+            state.Current.ReturnValue = GetCreatorDelegate(options)((Dictionary<string, TValue>)state.Current.ReturnValue!);
+        }
+
+        protected internal override bool OnWriteResume(Utf8JsonWriter writer, TCollection value, JsonSerializerOptions options, ref WriteStack state)
+        {
+            IEnumerator<KeyValuePair<string, TValue>> enumerator;
+            if (state.Current.CollectionEnumerator == null)
+            {
+                enumerator = value.GetEnumerator();
+                if (!enumerator.MoveNext())
+                {
+                    return true;
+                }
+            }
+            else
+            {
+                Debug.Assert(state.Current.CollectionEnumerator is Dictionary<string, TValue>.Enumerator);
+                enumerator = (Dictionary<string, TValue>.Enumerator)state.Current.CollectionEnumerator;
+            }
+
+            JsonConverter<TValue> converter = GetValueConverter(ref state);
+            do
+            {
+                if (ShouldFlush(writer, ref state))
```

Comment:
Just a reminder, a test should have failed since we didn't have `ShouldFlush` here. @layomia - let's make sure it is tested (such that if we were to remove this call, a test will fail).